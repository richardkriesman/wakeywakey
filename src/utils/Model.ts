import { AppDatabase } from "./AppDatabase";

export abstract class Model {

    protected db: AppDatabase;

    protected constructor(db: AppDatabase) {
        this.db = db;
    }

}
