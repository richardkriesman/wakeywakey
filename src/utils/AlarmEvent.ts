import { Alarm } from "../models/Alarm";

/**
 * The type of {@link AlarmEventType} being fired
 */
export enum AlarmEventType {
    Sleep,
    Wake,
    GetUp
}

export class AlarmEvent {

    public readonly alarm: Alarm;
    public readonly type: AlarmEventType;

    private readonly snoozeFn: (event: AlarmEvent) => Promise<void>;

    public constructor(alarm: Alarm, type: AlarmEventType, snoozeFn: (event: AlarmEvent) => Promise<void>) {
        this.alarm = alarm;
        this.snoozeFn = snoozeFn;
        this.type = type;
    }

    /**
     * Schedules the {@link AlarmEvent} to be replayed after its specified snooze time.
     */
    public snooze(): Promise<void> {
        return this.snoozeFn(this);
    }

}
