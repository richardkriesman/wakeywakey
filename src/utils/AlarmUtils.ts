import { AlarmModel } from "../models/AlarmModel";

export function getAlarmTitle(alarm: AlarmModel): string {
    const start = formatTime(alarm.sleepTime);
    const end = formatTime(alarm.getUpTime);
    return `${start} - ${end}`;
}

export function getAlarmSubtitle(alarm: AlarmModel): string {
    return alarm.days.join(", ");
}

export function formatTime(date: Date): string {
    return `${padTime(date.getHours())}:${padTime(date.getMinutes())}`;
}

export function padTime(time: number): string {
    return ("0" + time).slice(-2);
}

const AlarmUtils = {
    formatTime,
    getAlarmSubtitle,
    getAlarmTitle,
    padTime
};

export default AlarmUtils;
