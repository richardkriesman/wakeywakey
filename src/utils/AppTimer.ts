/**
 * @module utils
 */

/**
 * A function to execute when a {@link TimerEvent} fires.
 */
export type TimerHandler = (d?: Date) => void;

/**
 * A map to hold a list of {@link TimerHandler} functions to run when each {@link TimerEvent} fires.
 */
export type TimerHandlerMap = Map<string, TimerHandler[]>;

/**
 * Various timer-related events that can be listened for.
 */
export enum TimerEvent {
    START = "start",
    SECOND = "second",
    MINUTE = "minute",
    HOUR = "hour",
    STOP = "stop"
}

/**
 * String representation of TimerEvents. Too verbose?
 * Prevents attaching to events that will never be fired.
 */
export type TimerEventStr = "start" | "second" | "minute" | "hour" | "stop";

/**
 * Timer service: fires {@link TimerEvent}s when appropriate.
 *
 * Register a {@link TimerHandler} to listen for {@link TimerEvent}s by using {@link on}
 */
export class AppTimer {

    private static readonly INSTANCE: AppTimer = new AppTimer();

    /**
     * Retrieve the static instance of {@link AppTimer}.
     * @constructor
     */
    public static get Instance(): AppTimer {
        return AppTimer.INSTANCE;
    }

    /**
     * An instance field containing this service's {@link NodeJS.Timeout}'s ID.
     *
     * Can be undefined if no {@link NodeJS.Timeout} is scheduled for this {@TimerService} instance.
     */
    private timeout?: number = undefined;

    /**
     * The {@link Date} object representing the last time a timer tick happened.
     *
     * Updated on each tick.
     */
    private lastTick?: Date = undefined;

    /**
     * Getter for {@link NodeJS.Timeout} ID.
     *
     * Can be undefined if no {@link NodeJS.Timeout} is scheduled for this {@TimerService} instance.
     */
    public get timeoutId(): number {
        return this.timeout;
    }

    /**
     * A map to register {@link TimerEvent}s and their respective lists of {@link TimerHandler}s.
     */
    private handlers: TimerHandlerMap = new Map<TimerEvent, TimerHandler[]>();

    /**
     * Add a handler that listens for the given Timer event.
     * @param eventType A {@link TimerEvent} corresponding with the event to listen for
     * @param handler The {@link TimerHandler} to run when this event fires
     */
    public on(eventType: TimerEvent | TimerEventStr, handler: TimerHandler): void {
        let handlersArr: TimerHandler[] = this.handlers.get(eventType);
        handlersArr = handlersArr ? handlersArr : [];

        handlersArr.push(handler);

        this.handlers.set(eventType, handlersArr);
    }

    /**
     * Start this {@link AppTimer} instance.
     */
    public start() {
        console.log("starting AppTimer");
        this.fireAll(TimerEvent.START);
        this.step();
    }

    /**
     * Stop this {@link AppTimer} instance and clear its {@link NodeJS.Timeout}, if defined.
     */
    public stop() {
        console.log("stopping AppTimer");
        if (this.timeout) {
            clearTimeout(this.timeout);
            this.timeout = undefined;
        }

        this.fireAll(TimerEvent.STOP);
    }

    /**
     * Run all {@link TimerHandler} functions registered to this {@link TimerEvent}.
     * @param event The {@link TimerEvent} to fire all events for
     * @param date The {@link Date} to pass along to each handler
     */
    private fireAll(event: TimerEvent, date?: Date) {
        const handlersArr = this.handlers.get(event);

        // bail out if no handlers are registered
        if (!handlersArr) {
            return;
        }

        handlersArr.forEach((h: TimerHandler) => {
            h(date);
        });
    }

    /**
     * Handles a tick and schedules the next one using {@link scheduleNext}.
     */
    private step(fire?: boolean): void {
        const now: Date = new Date();

        if (fire && this.lastTick) {
            // fire events for this second
            this.fireAll(TimerEvent.SECOND, now);

            // fire minute event if minute has rolled over
            if (now.getMinutes() !== this.lastTick.getMinutes()) {
                this.fireAll(TimerEvent.MINUTE, now);
            }

            // fire hour event if hour has rolled over
            if (now.getHours() !== this.lastTick.getHours()) {
                this.fireAll(TimerEvent.HOUR, now);
            }
        }

        // update lastTick and schedule next step
        this.lastTick = now;
        this.scheduleNext();
    }

    /**
     * Schedule the next tick to fire when the next second rolls over.
     */
    private scheduleNext(): void {
        const now = new Date();
        const msUntilNextSecond: number = Math.min(1000, 1000 - now.getMilliseconds());
        this.timeout = setTimeout(this.step.bind(this, true), msUntilNextSecond);
    }

}

export default (() => AppTimer.Instance)();
