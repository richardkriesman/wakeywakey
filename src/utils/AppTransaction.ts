/**
 * A database handle that executes all queries in the context of a transaction
 */
export class AppTransaction {

    private transaction: SQLTransaction;

    public constructor(transaction: SQLTransaction) {
        this.transaction = transaction;
    }

    /**
     * Executes a SQL query, returning a {@link Promise} that resolves to a {@link SQLResultSet}.
     *
     * @param sql The SQL query to execute
     * @param args An array of values to substitute for "?" placeholders in the SQL statement.
     */
    public execute(sql: DOMString, args?: ObjectArray): Promise<SQLResultSet> {
        return new Promise<SQLResultSet>((resolve, reject) => {
            this.transaction.executeSql(sql, args ? args : [], (transaction, resultSet) => {
                resolve(resultSet);
            }, (transaction, error) => {
                reject(error);
                return true;
            });
        });
    }

}
