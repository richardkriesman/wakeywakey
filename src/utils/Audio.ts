import { Audio } from "expo";

export enum AlarmAudio {
    MusicBox = 0,
    Birds = 1,
    PagerBeeps = 2,
    Computer = 3,
    LoudAlarm = 4,
    NormalAlarm = 5,
    Special = 6
}

export async function getAlarmSound(audio: AlarmAudio): Promise<Audio.Sound> {
    const sound: Audio.Sound = new Audio.Sound();
    switch (audio) {
        case AlarmAudio.MusicBox:
            await sound.loadAsync(require("../../assets/audio/MusicBox.mp3"));
            break;
        case AlarmAudio.Birds:
            await sound.loadAsync(require("../../assets/audio/Birds.mp3"));
            break;
        case AlarmAudio.PagerBeeps:
            await sound.loadAsync(require("../../assets/audio/Pager.mp3"));
            break;
        case AlarmAudio.Computer:
            await sound.loadAsync(require("../../assets/audio/Computer.mp3"));
            break;
        case AlarmAudio.LoudAlarm:
            await sound.loadAsync(require("../../assets/audio/LoudAlarm.mp3"));
            break;
        case AlarmAudio.NormalAlarm:
            await sound.loadAsync(require("../../assets/audio/NormalAlarm.mp3"));
            break;
        case AlarmAudio.Special:
            await sound.loadAsync(require("../../assets/audio/Special.mp3"));
            break;
        default:
            throw new Error(`Audio ${audio} does not exist`);
    }
    return sound;
}
