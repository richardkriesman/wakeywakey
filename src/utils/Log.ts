/**
 * @module utils
 */

const MAX_TAG_SIZE: number = 10;

enum LogSeverity {
    Debug,
    Info,
    Warning,
    Error,
    Critical
}

enum BackgroundColor {
    Blue = "\u001b[44;1m",
    Green = "\u001b[42;1m",
    Red = "\u001b[41;1m",
    Reset = "\u001b[0m",
    White = "\u001b[47m",
    Yellow = "\u001b[43;1m"
}

enum TextColor {
    Black = "\u001b[30;1m",
    Red = "\u001b[31;1m",
    Yellow = "\u001b[33;1m"
}

export function debug(tag: string, message: string|Error): void {
    print(tag, LogSeverity.Debug, message);
}

export function info(tag: string, message: string|Error): void {
    print(tag, LogSeverity.Info, message);
}

export function warning(tag: string, message: string|Error): void {
    print(tag, LogSeverity.Warning, message);
}

export function error(tag: string, message: string|Error): void {
    print(tag, LogSeverity.Error, message);
}

export function critical(tag: string, message: string|Error): void {
    print(tag, LogSeverity.Critical, message);
}

function print(tag: string, severity: LogSeverity, message: string|Error): void {

    // format the message for display
    if (message instanceof Error) { // if this message is an error, display it as a string
        message = `${message.name}: ${message.message}\n${message.stack}`;
    } else {
        message = message.trim();
    }

    // add tag to message
    const paddingLen: number = (MAX_TAG_SIZE / 2) -
        Math.floor((tag.length <= MAX_TAG_SIZE ? tag.length : MAX_TAG_SIZE) / 2);
    for (let i = 0; i < paddingLen; i++) { // add padding for left side
        tag = ` ${tag}`;
    }
    for (let i = 0; i < paddingLen; i++) { // add padding for right side
        tag = `${tag} `;
    }
    if (tag.length % 2 !== 1) { // odd-length tag, add 1 extra space to the right
        tag += " ";
    }
    if (tag.length > MAX_TAG_SIZE) { // tag exceeds space limit, truncate it
        tag = tag.substring(0, MAX_TAG_SIZE);
    }
    tag = `${BackgroundColor.Green} ${tag} ${BackgroundColor.Reset}`;

    // create severity symbol
    let symbol: string;
    switch (severity) {
        case LogSeverity.Debug:
            symbol = `${BackgroundColor.White}${TextColor.Black} λ ${BackgroundColor.Reset}`;
            break;
        case LogSeverity.Info:
            symbol = `${BackgroundColor.Blue} i ${BackgroundColor.Reset}`;
            break;
        case LogSeverity.Warning:
            symbol = `${BackgroundColor.Yellow} w ${BackgroundColor.Reset}`;
            break;
        case LogSeverity.Error:
            symbol = `${BackgroundColor.Red} ! ${BackgroundColor.Reset}`;
            break;
        case LogSeverity.Critical:
            symbol = `${BackgroundColor.Red} ‼ ${BackgroundColor.Reset}`;
    }

    // add background to message
    switch (severity) {
        case LogSeverity.Info:
            message = `${message}`;
            break;
        case LogSeverity.Warning:
            message = `${TextColor.Yellow}${message}`;
            break;
        case LogSeverity.Error:
            message = `${TextColor.Red}${message}`;
            break;
        case LogSeverity.Critical:
            message = `${BackgroundColor.Red}${message}`;
    }

    // print message to console
    console.log(`${tag}${symbol}\t${message}${BackgroundColor.Reset}`);
}
