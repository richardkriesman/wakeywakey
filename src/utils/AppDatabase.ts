/**
 * @module utils
 */

import {FileSystem, SQLite} from "expo";
import * as Log from "./Log";

const DATABASE_NAME: string = "app";
const DATABASE_LOG_TAG: string = "Database";

export class AppDatabase {

    public static get initCount(): number {
        return this._initCount;
    }

    /**
     * Creates and opens a connection to the app's SQLite database, creating any tables that don't exist.
     */
    public static async init(): Promise<AppDatabase> {
        const appDb = new AppDatabase();

        // create preferences table
        if (!(await appDb.doesTableExist("preferences"))) {
            console.debug(`Database table "preferences" does not exist, creating it`);
            await appDb.execute(`
                CREATE TABLE preferences
                (
                    name VARCHAR not null
                        constraint preferences_pk primary key,
                    value VARCHAR not null
                )
            `);
        }

        AppDatabase._initCount++;
        return appDb;
    }

    /**
     * Tracks the number of database initializations. Used to catch accidental re-initializations when the connection
     * is not properly passed between screens.
     */
    private static _initCount: number = 0;

    private db: Database;

    private constructor() {
        this.db = SQLite.openDatabase(DATABASE_NAME);
        Log.info(DATABASE_LOG_TAG, `Opened database at ${FileSystem.documentDirectory}SQLite/${DATABASE_NAME}`);
    }

    /**
     * Determines whether a table with the given name exists in the database.
     *
     * @param name The table name
     */
    public async doesTableExist(name: string): Promise<boolean> {
        const result: SQLResultSet = await this.execute(`
            SELECT *
            FROM sqlite_master
            WHERE
                tbl_name = ?
        `, [name]);
        return result.rows.length > 0;
    }

    /**
     * Returns the value from the key/value table. Returns `null` if no such value exists.
     *
     * Note: This should only be used for storing general preference data that would not be appropriate in a relational
     * schema. Do not use the key/value table for storing models.
     *
     * @param key Unique key identifying the key/value pair.
     */
    public async getPreference(key: string): Promise<string|null> {
        const result: SQLResultSet = await this.execute(`
            SELECT *
            FROM preferences
            WHERE
                name = ?
        `, [key]);
        if (result.rows.length > 0) {
            const value: string = result.rows.item(0).value;
            Log.info(DATABASE_LOG_TAG, `GET ${key}: ${value}`);
            return value;
        } else {
            Log.info(DATABASE_LOG_TAG, `GET ${key}: (null)`);
            return null;
        }
    }

    /**
     * Executes a SQL query against the database, returning a {@link SQLResultSet} containing the resultant rows.
     *
     * @param sql The SQL query to execute.
     * @param args An array of values to substitute for "?" placeholders in the SQL statement.
     */
    public execute(sql: DOMString, args: ObjectArray = []): Promise<SQLResultSet> {
        return new Promise<SQLResultSet>((resolve, reject) => {
            this.db.transaction((transaction) => { // wrap the query in a transaction
                transaction.executeSql(sql, args, (_, resultSet) => { // success, return result set
                    resolve(resultSet);
                }, (_, error) => { // error, reject
                    console.log(error);
                    reject(error);
                    return true;
                });
            });
        });
    }

    /**
     * Sets the value for a key in the key/value table. If a value already exists, it will be overwritten.
     *
     * Note: This should only be used for storing general preference data that would not be appropriate in a relational
     * schema. Do not use the key/value table for storing models.
     *
     * @param key Unique key
     * @param value Value to set for the key
     */
    public async setPreference(key: string, value: string): Promise<void> {
        await this.execute(`
            INSERT OR REPLACE INTO preferences
                (name, value)
            VALUES
                (?, ?);
        `, [key, value]);
        Log.info(DATABASE_LOG_TAG, `SET ${key}: ${value}`);
    }

}
