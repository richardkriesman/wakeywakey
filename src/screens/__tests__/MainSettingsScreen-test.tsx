import React from "react";
import "react-native";
import renderer from "react-test-renderer";
import {TestEnvironment} from "../../utils/testing";
import SchedulesListScreen from "../settings/SchedulesListScreen";

let env: TestEnvironment;
beforeEach((done) => {
    TestEnvironment.init()
        .then((newEnv) => {
            env = newEnv;
            done();
        });
});

it("renders the screen", async () => {
    const tree = renderer.create(
        <SchedulesListScreen navigation={env.navigationProp} />
    ).toJSON();
    expect(tree).toMatchSnapshot();
});
