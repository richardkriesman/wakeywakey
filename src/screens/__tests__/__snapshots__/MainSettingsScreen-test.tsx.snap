// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App snapshot renders the screen 1`] = `
<RCTSafeAreaView
  style={
    Object {
      "flex": 1,
    }
  }
/>
`;
